libs/vkd3d/shared_fence.c

```diff
@@ -0,0 +1,149 @@
+#include "shared_fence.h"
+
+/* For now the implementation isn't cross-process.
+ * When fixing this, we just need to reimplement the blocking wait with an event handle,
+ * removing the need for the conditional variable, and use a win32 mutex instead of a critical section */
+
+#define MAX_WAIT_COUNT 100
+
+struct waiter_thread
+{
+    bool waiting, signaled;
+    HANDLE handle;
+    UINT64 value;
+};
+
+struct shared_fence
+{
+    DWORD pid;
+    UINT64 value;
+    CRITICAL_SECTION *cs;
+    CONDITION_VARIABLE *cv;
+    struct waiter_thread waiters[MAX_WAIT_COUNT];
+};
+
+HANDLE create_shared_fence(UINT64 initial_value)
+{
+    HANDLE mapping = CreateFileMappingW(INVALID_HANDLE_VALUE, NULL, PAGE_READWRITE | SEC_COMMIT, 0, sizeof(struct shared_fence), NULL);
+    struct shared_fence *fence = open_shared_fence(mapping);
+
+    fence->pid = GetCurrentProcessId();
+    fence->value = initial_value;
+    fence->cs = malloc(sizeof(*fence->cs));
+    fence->cv = malloc(sizeof(*fence->cv));
+    InitializeCriticalSection(fence->cs);
+    InitializeConditionVariable(fence->cv);
+
+    shared_fence_close(fence);
+
+    return mapping;
+}
+
+struct shared_fence *open_shared_fence(HANDLE handle)
+{
+    return MapViewOfFile(handle, FILE_MAP_WRITE, 0, 0, sizeof(struct shared_fence));
+}
+
+void shared_fence_close(struct shared_fence *fence)
+{
+    UnmapViewOfFile(fence);
+}
+
+UINT64 shared_fence_get_value(struct shared_fence *fence)
+{
+    UINT64 value;
+    EnterCriticalSection(fence->cs);
+    value = fence->value;
+    LeaveCriticalSection(fence->cs);
+    return value;
+}
+
+/* Equivalent of SetEventOnCompletion*/
+bool shared_fence_wait(struct shared_fence *fence, HANDLE handle, UINT64 value)
+{
+    struct waiter_thread *waiter;
+    unsigned int i;
+
+    if (fence->pid != GetCurrentProcessId())
+        return false;
+
+    EnterCriticalSection(fence->cs);
+
+    if (fence->value >= value)
+    {
+        LeaveCriticalSection(fence->cs);
+
+        if (handle != NULL && handle != INVALID_HANDLE_VALUE)
+            SetEvent(handle);
+
+        return true;
+    }
+
+    for (i = 0; i < MAX_WAIT_COUNT; i++)
+    {
+        waiter = &fence->waiters[i];
+
+        if (!waiter->waiting)
+        {
+            waiter->waiting = true;
+            waiter->signaled = false;
+            waiter->handle = handle;
+            waiter->value = value;
+            break;
+        }
+    }
+
+    if (i == MAX_WAIT_COUNT)
+    {
+        LeaveCriticalSection(fence->cs);
+        return false;
+    }
+
+    if (handle != NULL && handle != INVALID_HANDLE_VALUE)
+    {
+        LeaveCriticalSection(fence->cs);
+        return true;
+    }
+
+    while (!waiter->signaled)
```

Comment:
Is this the path for when we have SetEventOnCompletion(NULL) which is blocking? Would be easier to understand the control flow if this was commented.